adam-core/src/main/scala/org/bdgenomics/adam/rdd/GenomicRDD.scala

```diff
@@ -725,45 +1207,151 @@ trait GenomicRDD[T, U <: GenomicRDD[T, U]] {
+   */
+  def coPartitionByGenomicRegion[X, Y <: GenomicRDD[X, Y]](rddToCoPartitionWith: GenomicRDD[X, Y])(implicit tTag: ClassTag[T], xTag: ClassTag[X]): U = {
+    // if the other RDD is not sorted, we can't guarantee copartition
+    assert(rddToCoPartitionWith.isSorted)
+    // materialize the partition map if it's not already in memory
+    val collectedDestinationPartitionMap = rddToCoPartitionWith.partitionMap
+    //number of partitions
+    val partitions = collectedDestinationPartitionMap.length
+    // We end up with an RDD that has the same partition map as the rddToCoPartitionWith
+    val finalPartitionedRDD = flattenRddByRegions.mapPartitionsWithIndex((idx, iter) => {
+      val listRepresentation = iter.map(_.swap).toList
+      // key this by the destination partition index
+      // we end up with a (k, v) pair for each partition such that the key is the 
+      // destination partition index and the value is the list of data that will be 
+      // put in that partition. The data is still unsorted at this point
+      for (partition <- collectedDestinationPartitionMap.zipWithIndex) yield {
+        (partition._2, listRepresentation.filter(f => (f._2.start >= partition._1._1.start && f._2.start <= partition._1._2.end) ||
+          (f._2.end >= partition._1._1.start && f._2.end <= partition._1._2.end)))
+      }
+    }.toIterator.filter(_._2.nonEmpty), preservesPartitioning = true)
+      .partitionBy(new GenomicPositionRangePartitioner(partitions))
+      // now we sort all the data on a partition
+      // it is more efficient this way because we avoid a shuffle operation
+      // and we have to go through all the data anyway
+      .mapPartitions(iter => {
+        iter.flatMap(_._2).toList.sortBy(_._2).toIterator
+      }, preservesPartitioning = true)
+
+    // here we get the new partition map
+    // for now, we can't use the left RDD's partition map because
+    // the bounds don't always include the values we copartitioned with.
+    // in other words, joins allow data that starts before a given
+    // ReferenceRegion to combine given they overlap.
+    // this means that we cannot use the bounds of the left RDD because
+    // they may not neccessarily be the bounds of the right RDD
+    val newPartitionMap = finalPartitionedRDD.mapPartitions(iter => {
+      getPartitionMapBoundsFromRdd(iter.map(_.swap))
+    }, preservesPartitioning = true)
+
+    replaceRdd(finalPartitionedRDD.keys, Some(newPartitionMap))
+  }
+
+  /**
+    * Gets the partition bounds from a ReferenceRegion keyed Iterator
+    *
+    * @param iter The data on a given partition. ReferenceRegion keyed
+    * @return The bounds of the ReferenceRegions on that partition, in an Iterator
+    */
+  private def getPartitionMapBoundsFromRdd(iter: Iterator[(ReferenceRegion, T)]): Iterator[(ReferenceRegion, ReferenceRegion)] = {
+    if (iter.isEmpty) Iterator()
```

Comment:
In general, I'm anti dropping the braces on if-else clauses. Also, this won't surprise anyone who's familiar with my style and general way of approaching Scala, but I'd generally prefer this function written as a tailrec call:

```
@tailrec private def getBounds(iter: Iterator[(ReferenceRegion, T)],
  optStart: Option[ReferenceRegion] = None,
  optEnd: Option[ReferenceRegion] = None) {
  if (iter.hasNext) {
    (optStart, optEnd) match {
       case (None, None) => Iterator.empty
       case (Some(startRegion), None) => Iterator((startRegion, startRegion))
       case (Some(startRegion), Some(endRegion)) => Iterator((startRegion, endRegion))
       case _ => throw new IllegalStateException("Detailed error message here")
    }
  } else {
    val (newOptStart, newOptEnd) = optStart.fold((Some(iter.next), None))
      (currOptStart => (currOptStart, Some(iter.next))
    getBounds(iter, newOptStart, newOptEnd)
  }
```

The `@tailrec` annotation is at `scala.annotation.tailrec`